crates/sound/src/device.rs

```diff
@@ -70,49 +76,301 @@ impl VhostUserSoundThread {
+                vring.disable_notification().unwrap();
+                match queue_idx {
+                    CONTROL_QUEUE_IDX => self.process_control(vring, audio_backend),
+                    EVENT_QUEUE_IDX => self.process_event(vring),
+                    TX_QUEUE_IDX => self.process_tx(vring, audio_backend),
+                    RX_QUEUE_IDX => self.process_rx(vring, audio_backend),
+                    _ => Err(Error::HandleUnknownEvent.into()),
+                }?;
+                if !vring.enable_notification().unwrap() {
+                    break;
+                }
+            }
+        } else {
+            // Without EVENT_IDX, a single call is enough.
+            match queue_idx {
+                CONTROL_QUEUE_IDX => self.process_control(vring, audio_backend),
+                EVENT_QUEUE_IDX => self.process_event(vring),
+                TX_QUEUE_IDX => self.process_tx(vring, audio_backend),
+                RX_QUEUE_IDX => self.process_rx(vring, audio_backend),
+                _ => Err(Error::HandleUnknownEvent.into()),
+            }?;
+        }
+        Ok(false)
     }
 
-    fn process_control(&self, _vring: &VringRwLock) -> IoResult<bool> {
-        let requests: Vec<_> = _vring
+    fn process_control(
+        &self,
+        vring: &VringRwLock,
+        audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
+    ) -> IoResult<bool> {
+        let requests: Vec<SoundDescriptorChain> = vring
             .get_mut()
             .get_queue_mut()
             .iter(self.mem.as_ref().unwrap().memory())
             .map_err(|_| Error::DescriptorNotFound)?
             .collect();
-        dbg!(&requests);
 
-        Ok(true)
+        if requests.is_empty() {
+            return Ok(true);
+        }
+
+        // Reply to some requests right away, and defer others to the audio backend (for
+        // example PcmRelease needs to complete all I/O before replying.
+        //
+        // Mark `any` as true if we need to reply to any request right away so we can
+        // signal the queue as used.
+        let mut any = false;
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+            // Request descriptor.
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0).into());
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioSoundHeader>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            // Keep track of bytes that will be written in the VQ.
+            let mut len = 0;
+
+            // Reply header descriptor.
+            let desc_hdr = descriptors[1];
+            if !desc_hdr.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1).into());
+            }
+            len += desc_hdr.len();
```

Comment:
I would move this when we really write the descriptor, just to be more readable.

Reply:
+1 will do.